### PR TITLE
GT-631 : update tool info home design (Phase 1)

### PR DESCRIPTION
**Pull request by gyasistory**
This is just changing the UI to reflect the design present in Figma.

https://jira.cru.org/browse/GT-631

### CODE REVIEW

File: ui/base/src/main/res/values/styles_widgets.xml [hunk 1]
```diff
@@ -19,6 +19,23 @@
...
         <item name="backgroundTint">?attr/colorButtonNormal</item>
     </style>
+    <style name="Widget.GodTools.Button.Outline.Red" parent="Widget.GodTools.Button.Outline">
+        <item name="backgroundTint">@color/gt_red</item>
+        <item name="android:textColor">@color/gt_red</item>
+    </style>
+    <style name="Widget.GodTools.Button.Outline.Blue" parent="Widget.GodTools.Button.Outline">
+        <item name="backgroundTint">@color/gt_blue</item>
+    </style>
+    <style name="Widget.GodTools.Button.Borderless.Colored" parent="Widget.AppCompat.Button.Borderless">
```
Comment (frett) [inline ui/base/src/main/res/values/styles_widgets.xml]: what's the difference between `Borderless` and the regular `Colored` buttons?

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java [hunk 2]
```diff
@@ -280,24 +288,23 @@ private void updateViews() {
...
-                                           .map(l -> LocaleUtils.getDisplayName(l, mLanguagesView.getContext(), null,
-                                                                                null))
-                                           .withoutNulls()
-                                           .sorted(String.CASE_INSENSITIVE_ORDER)
-                                           .reduce((l1, l2) -> l1 + ", " + l2)
-                                           .orElse(""));
+
+        final int count = mLanguages.size();
+        if (mViewPager != null) {
+            ToolsDetailAdapter adapter = new ToolsDetailAdapter();
```
Comment (frett) [inline app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: you didn't include the adapter and it should probably be named `ToolDetailsAdapter`

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java [hunk 3]
```diff
@@ -280,24 +288,23 @@ private void updateViews() {
...
+            mViewPager.setAdapter(adapter);
+            if (mTabLayout != null) {
+                mTabLayout.setupWithViewPager(mViewPager, true);
+                if (mTabLayout.getTabAt(1) != null) {
+                    mTabLayout.getTabAt(1).setText(requireContext().getResources()
+                            .getQuantityString(R.plurals.label_tools_languages, count, count));
+                }
+                if (mTabLayout.getTabAt(0) != null) {
+                    mTabLayout.getTabAt(0).setText(R.string.label_tools_about);
```
Comment (frett) [inline app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: why can't you just set this text directly in the XML?

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java [hunk 4]
```diff
@@ -332,6 +344,30 @@ void removeTool() {
...
+                            .getLatestPublishedManifest(mTool.getCode(), primaryLanguage);
+
+                    TractActivity.start(requireActivity(), mTool.getCode(), primaryLanguage,
+                                        parallelLanguages);
+                    break;
+                case ARTICLE:
+                    CategoriesActivityKt.startCategoriesActivity(requireActivity(), mTool.getCode(),
+                                                                 primaryLanguage);
+                    break;
+            }
```
Comment (frett) [inline app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: I'm not a huge fan of duplicating this logic here, what are the options available to be able to re-use this between the main tool list and this fragment?

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java (hunk 3 above)
Comment (gyasistory) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: Android_Emulator_-_Pixel_3a_API_29_5554

This is what I get when I try to set the text in the XML.

File: ui/base/src/main/res/values/styles_widgets.xml (hunk 1 above)
Comment (gyasistory) [reply ui/base/src/main/res/values/styles_widgets.xml]: @frett I created this style to match the size and structure of the Outline style.  This way they look similar to each other.

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java (hunk 2 above)
Comment (gyasistory) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: @frett not sure what you mean by I didn't include the adapter?

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java (hunk 4 above)
Comment (gyasistory) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: OK, I will look to see if I can find a way.

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java [hunk 5]
```diff
@@ -280,24 +286,23 @@ private void updateViews() {
...
-                                           .map(l -> LocaleUtils.getDisplayName(l, mLanguagesView.getContext(), null,
-                                                                                null))
-                                           .withoutNulls()
-                                           .sorted(String.CASE_INSENSITIVE_ORDER)
-                                           .reduce((l1, l2) -> l1 + ", " + l2)
-                                           .orElse(""));
+
+        final int count = mLanguages.size();
+        if (mViewPager != null) {
+            ToolDetailsAdapter adapter = new ToolDetailsAdapter();
```
Comment (frett) [inline app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: `updateViews()` is called whenever backing data changes. creating a brand new Adapter every time data changes is heavy. `updateViews()` should only be manipulating existing views, not recreating a significant part of the view hierarchy.

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java (hunk 3 above)
Comment (frett) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: You should leverage `PagerAdapter.getTitle()` to drive the tab titles, it's what the method is actually for

Comment (wrandall22) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: @gyasistory did you look into using the `getPageTitle` method as Daniel suggested?

Comment (wrandall22) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: Looks like you did, but just making sure.

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -45,6 +45,7 @@
     </plurals>
     <string name="text_tools_all_installed">You have downloaded all available tools.</string>
     <string name="menu_add_to_home">Add to Home screen</string>
+    <string name="action_tools_open">open tool</string>
```
Comment (wrandall22) [inline app/src/main/res/values/strings.xml]: Should this be `Open Tool` or does case not matter?

File: app/src/main/java/org/cru/godtools/util/ActivityUtils.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    // launch activity based on the tool type
+    when (type) {
+        Type.TRACT -> {
+            // start pre-loading the tract in the first language
+            ManifestManager.getInstance(context).getLatestPublishedManifest(code, languages[0])
+            TractActivity.start(context, code, *languages)
+        }
+        Type.ARTICLE -> context.startCategoriesActivity(code, languages[0])
+    }
+}
```
Comment (wrandall22) [inline app/src/main/java/org/cru/godtools/util/ActivityUtils.kt]: How much does this project care about newline at the end of files?

Comment (frett) [reply app/src/main/java/org/cru/godtools/util/ActivityUtils.kt]: I personally try to enforce a new line on the last line. I've enabled some ktlint/checkstyle rules to enforce it in certain places, but not everywhere evidently

File: app/src/main/res/values/strings.xml (hunk 6 above)
Comment (frett) [reply app/src/main/res/values/strings.xml]: Ann already changed this upstream, next time we sync translations from OneSky this should be fixed

Comment (frett) [reply app/src/main/res/values/strings.xml]: my 1 minor nit for this line is to move it up with the other actions for the tools UI (lines 35-36)

File: app/src/main/java/org/cru/godtools/util/ActivityUtils.kt [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+fun openToolActivity(context: Activity, code: String, type: Type, vararg languages: Locale) {
+    // launch activity based on the tool type
+    when (type) {
+        Type.TRACT -> {
+            // start pre-loading the tract in the first language
+            ManifestManager.getInstance(context).getLatestPublishedManifest(code, languages[0])
+            TractActivity.start(context, code, *languages)
+        }
+        Type.ARTICLE -> context.startCategoriesActivity(code, languages[0])
```
Comment (wrandall22) [inline app/src/main/java/org/cru/godtools/util/ActivityUtils.kt]: I'm a bit curious how this:
`context.startCategoriesActivity(code, languages[0])`
replaces this successfully:
`CategoriesActivityKt.startCategoriesActivity(this, code, languages[0]);`

Comment (frett) [reply app/src/main/java/org/cru/godtools/util/ActivityUtils.kt]: Kotlin extension functions https://kotlinlang.org/docs/reference/extensions.html#extension-functions

when Kotlin compiles, it replaces it with a static method call in the generated bytecode

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java (hunk 2 above)
Comment (frett) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: I hadn't noticed it was an inner class, I was looking for a separate `ToolDetailsAdapter` source file.

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java [hunk 9]
```diff
@@ -275,29 +281,13 @@ private void updatePinShortcutAction() {
     }
 
     private void updateViews() {
+        setUpViewPager();
```
Comment (frett) [inline app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: you are still setting up the view pager every time the view is being updated

Comment (frett) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: this should only be called when the view is created, not every time `updateViews()` is called

Comment (gyasistory) [reply app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: @frett I was unable to remove the setAdapter from the `updateViews`.  I tried doing a notifyDataSet Change and InitiateItem but I wasn't able to see the text in the view pager.

File: app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java [hunk 10]
```diff
@@ -477,28 +480,29 @@ private void setUpViewPager() {
         @NonNull
         @Override
         public Object instantiateItem(@NonNull ViewGroup container, int position) {
-            AppCompatTextView textView = (AppCompatTextView) LayoutInflater.from(container.getContext())
-                    .inflate(R.layout.tool_detail_text_view, container, false);
+            AppCompatTextView textView =
```
Comment (wrandall22) [inline app/src/main/java/org/cru/godtools/fragment/ToolDetailsFragment.java]: It would be nice if whitespace changes were done in a separate commit.

**Pull request merged**